data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/stream/StreamWorker.java

```diff
@@ -0,0 +1,164 @@
+        final String collectionDbName = streamPartition.getCollection();
+        List<String> collectionDBNameList = List.of(collectionDbName.split(COLLECTION_SPLITTER));
+        if (collectionDBNameList.size() < 2) {
+            throw new IllegalArgumentException("Invalid Collection Name. Must be in db.collection format");
+        }
+        int recordCount = 0;
+        final List<String> records = new ArrayList<>();
+        // TODO: create acknowledgementSet
+        AcknowledgementSet acknowledgementSet = null;
+        String checkPointToken = null;
+        try (MongoClient mongoClient = MongoDBConnection.getMongoClient(sourceConfig)) {
+            // Access the database
+            MongoDatabase database = mongoClient.getDatabase(collectionDBNameList.get(0));
+
+            // Access the collection you want to stream data from
+            MongoCollection<Document> collection = database.getCollection(collectionDBNameList.get(1));
+
+            try (MongoCursor<ChangeStreamDocument<Document>> cursor = getChangeStreamCursor(collection, resumeToken.orElse(null))) {
+                while (shouldWaitForExport(streamPartition) && !Thread.currentThread().isInterrupted()) {
+                    LOG.info("Initial load not complete for collection {}, waiting for initial lo be complete before resuming streams.", collectionDbName);
+                    try {
+                        Thread.sleep(DEFAULT_EXPORT_COMPLETE_WAIT_INTERVAL_MILLIS);
+                    } catch (final InterruptedException ex) {
+                        LOG.info("The StreamScheduler was interrupted while waiting to retry, stopping processing");
+                        break;
+                    }
+                }
+
+                while (cursor.hasNext() && !Thread.currentThread().isInterrupted()) {
+                    try {
+                        final ChangeStreamDocument<Document> document = cursor.next();
+                        final String record = document.getFullDocument().toJson(writerSettings);
+
+                        checkPointToken = document.getResumeToken().toJson(writerSettings);
+
+                        records.add(record);
+                        recordCount += 1;
+
+                        if (recordCount % defaultFlushBatchSize == 0) {
+                            LOG.debug("Write to buffer for line " + (recordCount - defaultFlushBatchSize) + " to " + recordCount);
+                            recordBufferWriter.writeToBuffer(acknowledgementSet, records);
+                            records.clear();
+                            partitionCheckpoint.checkpoint(checkPointToken, recordCount);
+                            LOG.debug("Perform regular checkpointing for stream Loader");
+                            partitionCheckpoint.checkpoint(checkPointToken, recordCount);
+                        }
+                        successItemsCounter.increment();
```

Comment:
Strictly speaking, it's not successfully written to buffer yet. Should we move it inside the if block after writeToBuffer() call?

Reply:
yeah, makes sense.